_posts/2022-10-28-DistributedJakartaEnterpriseBeanTimers.adoc

```diff
@@ -0,0 +1,115 @@
+---
+layout: post
+title: 
+date: 2022-10-28
+tags: wildfly ejb timers distributed clustered clustering infinispan
+synopsis: A new distributed Jakarta Enterprise Bean TimerService for WildFly
+author: pferraro
+---
+
+= Distributed Jakarta Enterprise Bean Timers
+
+The Jakarta Enterprise Bean Timer Service allows an application to schedule application callbacks for temporal events.
+
+WildFly 27 adds a new distributed TimerService implementation that improves the scalability and efficiency of Jakarta Enterprise Bean timers in a cluster by leveraging Infinispan for both timer availability/persistence and distribution of timer scheduling.
+
+In general, the distributed TimerService implementation has the following characteristics:
+
+* Persistent auto-timers are created once per cluster, in accordance with section 12.2.2 of the Jakarta Enterprise Bean specification.
+* Non-persistent auto-timers are created once per cluster member, in accordance with section 12.2.3 of the Jakarta Enterprise Bean specification.
+* Manually created persistent timers will initially schedule locally on the cluster member that created it.
+* Changes to cluster topology will initiate changes to the primary ownership of existing timers, such that timer execution will be shared relatively evenly by all cluster members on which the same Enterprise Java Bean is deployed.
+** This means that a new cluster member will assume responsibility for a share of existing timers, while timers previously owned by a leaving member will be reassigned to surviving cluster members.
+** Users can manipulate the nuances of timer distribution by modifying the number of segments owned by each cluster member (via Infinispan subsystem configuration).
+* `TimerService.getTimers()` returns all active timers associated with a given bean, including those created on, or currently scheduled by, other cluster members.
+* `TimerService.getAllTimers()` returns all active timers from all beans within a given module, including those created on, or currently scheduled by, other cluster members.
+
+== Configuring Distributed Jakarta Enterprise Bean Timers
+
+Distributed Jakarta Enterprise Bean timers are configured via the ejb3 subsystem.  The default "ha" and "full-ha" profiles are pre-configured to support distributed timers.
+e.g.
+
+[source,xml]
+----
+<subsystem xmlns="urn:jboss:domain:ejb3:10.0">
+	<!-- ... -->
+	<timer-service default-persistent-timer-management="persistent" default-transient-timer-management="transient"/>
+	<!-- ... -->
+</subsystem>
+----
+
+The `default-persistent-timer-management` and `default-transient-timer-management` that define the server-wide default timer management used for persistent and transient (i.e. non-persistent) timers referencing resources from the `distributable-ejb` subsystem (new to WildFly 27).
```

Comment:
I didn't understand this. Perhaps...

s/that define/attributes define/g

s/referencing/by referencing/g

Reply:
Oy. You are right. That sentence makes no sense - will fix.

Reply:
Fixed.